# Predicting the price of diamonds

## Table of contents <a id="contents"></a>
- [Section 1.0 (Introduction)](#intro)
- [Section 1.1 (Dataset source)](#dSource) Y (Could continue, will need to reference)
- [Section 1.2 (Dataset details)](#dDetails) - Discuss the style of dataset - Jacob Hill
- [Section 1.3 (Dataset variables)](#dVariables) - Francis Monaghan
- [Section 1.4 (Response variable)](#rVariable) - Francis Monaghan
- [Section 2.0 (Goals & Objectives)](#Objectives) - Francis Monaghan/Jacob Hill
- [Section 3.0 (Data Cleaning & Preprocessing)](#Preprocessing) - Need to look at missing values/outliers/incorrect values (Thomas Yao)
- [Section 4.0 (Data Exploration and Visualisation)](#Visualisation) - Thomas Yao
- [Section 4.5 (Literature Review)](#Review) - Looking at this on Friday
- [Section 5.0 (Summary & Conclusions)](#Conclusions)
- [Section 6.0 (References)](#References)

Other work:
Cleaning/Prepocessing, looking at https://github.com/vaksakalli/datasets/blob/master/diamonds_clean_5000.csv, comparing it to our dataset - Jacob Hill

In [1]:
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None) 

df = pd.read_csv("diamonds.csv")

###
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use("ggplot")
%matplotlib inline
%config InlineBackend.figure_format = "svg"
###

In [2]:
# This proves there are no not a number values in any observation, 
# however there may be missing values in a different format than null
for col in df.columns.values:
    if df[col].isna().sum() > 0:
        naValueSum = df[col].isna().sum()
        print(f"{col} has {missingValueSum} NaN values")

(53940, 10)

## Intro <a id='intro'></a>
## Dataset source <a id='dSource'></a>

Our Dataset was sourced from our lecturer's example datasets provided on GitHub at https://github.com/vaksakalli/datasets/blob/master/diamonds.csv 

In [3]:
df.columns.values

(53940, 10)

In [5]:
# This proves there are no not a number values in any observation, 
# however there may be missing values in a different format than null
for col in df.columns.values:
    if df[col].isna().sum() > 0:
        naValueSum = df[col].isna().sum()
        print(f"{col} has {missingValueSum} NaN values")

JH - could we show one of these instead/as well to be clear?

In [9]:
df.isna().sum()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
x          0
y          0
z          0
price      0
dtype: int64

In [10]:
 df.isnull().values.any()

False

In [6]:
# This is every unique value in each column, here we can analyse for bad values
for col in df.columns.values:
    print(f"{col} \n {df[col].unique()}")

carat 
 [0.23 0.21 0.29 0.31 0.24 0.26 0.22 0.3  0.2  0.32 0.33 0.25 0.35 0.42
 0.28 0.38 0.7  0.86 0.71 0.78 0.96 0.73 0.8  0.75 0.74 0.81 0.59 0.9
 0.91 0.61 0.77 0.63 0.76 0.64 0.72 0.79 0.58 1.17 0.6  0.83 0.54 0.98
 0.52 1.01 0.53 0.84 0.51 1.05 0.55 0.87 1.   0.57 0.82 1.04 0.93 1.2
 0.99 0.34 0.43 0.36 0.95 0.89 1.02 0.97 0.56 0.85 0.92 1.27 0.66 1.12
 0.68 1.03 0.62 1.22 1.08 0.88 0.5  1.19 0.39 0.65 1.24 1.5  0.27 0.41
 1.13 1.06 0.69 0.4  1.14 0.94 1.29 1.52 1.16 1.21 1.23 1.09 0.67 1.11
 1.1  1.18 1.15 1.25 1.07 1.28 1.51 0.37 1.31 1.26 1.39 1.44 1.35 1.3
 1.32 1.41 1.36 1.45 1.34 1.58 1.54 1.38 1.33 1.74 1.64 1.47 1.4  1.55
 1.95 2.   1.37 1.83 1.62 1.57 1.69 2.06 1.72 1.66 2.14 1.49 1.46 2.15
 1.96 2.22 1.7  1.53 1.85 2.01 2.27 1.68 1.56 1.81 1.65 1.82 2.03 1.73
 1.59 1.42 1.43 2.08 1.48 1.6  2.49 1.71 2.02 2.07 3.   2.21 2.1  1.91
 2.25 2.17 2.32 2.72 1.61 2.23 2.11 2.05 1.63 2.3  2.31 1.75 2.04 2.12
 1.77 2.5  1.8  1.67 1.84 2.2  3.01 1.88 2.33 2.68 2.34 1.9  2.16 2.74
 

Might be good to summaries the overivew of data something like below.

In [11]:
df.nunique()

carat        273
cut            5
color          7
clarity        8
depth        184
table        127
x            554
y            552
z            375
price      11602
dtype: int64

## Introduction <a id='intro'></a>

Return to [Table of contents](#contents)

## Dataset source <a id='dSource'></a>

Our Dataset was sourced from our lecturer's example datasets provided on GitHub at https://github.com/vaksakalli/datasets/blob/master/diamonds.csv 

Return to [Table of contents](#contents)

## Dataset details <a id='dDetails'></a>
Our dataset contains 53,940 observations, and 10 variables.

# Dataset Variables <a id='dVariables'></a>
The variables describe the physical features of each diamond observed, such as size, clarity.

In [27]:
inferenceCounter = 0
for col in df.columns.values:
    print(f"{col}\'s data type is {df[col].dtypes}")
    if df[col].dtypes != df[col].infer_objects():
        print(f"Whilst {col} is stored as {df[col].dtypes}, Pandas is inferring it as df[col].infer_objects")
        inferenceCounter += 1
if inferenceCounter == 0:
    print("\nPandas has inferred that all columns are the correct data type")

carat's data type is float64
cut's data type is object
color's data type is object
clarity's data type is object
depth's data type is float64
table's data type is float64
x's data type is float64
y's data type is float64
z's data type is float64
price's data type is int64

Pandas has inferred that all columns are the correct data type


In [4]:
df.columns.values

array(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'x', 'y',
       'z', 'price'], dtype=object)

# Response Variable <a id='dResponse'></a>
The response or target variable for this dataset is the price that the diamond was sold for.

# Goals & Objectives <a id='gObjectives'></a>
The goal for our data analysis is to find a pattern between physical attributes of the diamond and the resulting price the diamond sold for.

Return to [Table of contents](#contents)

## Dataset Variables <a id='dVariables'></a>
 Explain the variables in your dataset that you will be including in your project in a table format. To be clear, please include only the variables you will use in your project and leave out the ones you will not be using in your project. In your table, you need to have one variable per row with the following 4 columns:
 
    - Name of the variable
    - Data type (numeric, binary, nominal categorical, ordinal categorical, date)
    - Units ("Unknown" or "NA" for Not Applicable are acceptable)
    - Brief description

In [12]:
df.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
x          float64
y          float64
z          float64
price        int64
dtype: object

As we can see from the data types, cut color and clarity are categorical (ordinal?). The rest are numeric.

Return to [Table of contents](#contents)

## Response Variable <a id='rVariable'></a>
Our response variable will be the `price` of diamonds, which is numeric.

Return to [Table of contents](#contents)

## Goals & Objectives <a id='Objectives'></a>
(10 points) for modelling this particular data. 

Return to [Table of contents](#contents)

## Data Cleaning & Preprocessing <a id='Preprocessing'></a>
If your dataset is **already clean** and ready for modelling, you will get the full score for this task, our treat! For this section, you will need to add subsections as appropriate for better organisation of your work.

Return to [Table of contents](#contents)

## Data Exploration & Visualisation <a id='Visualisation'></a>
as appropriate with proper explanation: charts, graphs, boxplots, numerical summaries, etc. For this section, you will need to add subsections as appropriate for better organisation. 

Return to [Table of contents](#contents)

## Literature Review <a id='Review'></a>
(Advanced Submission Mode Only, 15 points) 

Return to [Table of contents](#contents)

## Summary & Conclusions <a id='Conclusions'></a>
of the first phase of your project: a comprehensive summary of Phase 1 and any insights you gained in this phase as they relate to your goals and objectives.

Return to [Table of contents](#contents)

## References <a id='References'></a>
For citing any book, article, paper, website, etc. in your report, you need to use the APA style

Return to [Table of contents](#contents)

## Comment - Francis
I just realised there is a clean dataset provided, would we prefer to use that, or demonstrate us cleaning the data?
It would be pretty easy to make sure our data is clean by comparing it to the clean dataset provided.